File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);  
        int n = nums.length;
        int l = 0;
        int r = n - 1;
        long count = 0;
        while (l < r) {
            //System.out.println(l + " " + r);
            if (nums[l] + nums[r] >= lower && nums[l] + nums[r] <= upper) {
                // binary search
                int L = l;
                int R = r;
                int mid = r;
                while (L <= R) {
                    mid = L + (R - L) / 2;
                    if (mid < r && nums[mid] + nums[r] <= upper) {
                        L = mid + 1;
                    } else {
                        R = mid - 1;
                    }
                }
                //System.out.println(L + " - " + l);
                count += (L - l);
                r--;
               
            } else if (nums[l] + nums[r] > upper) {
                r--;
            } else {
                l++;
            }
           
        }

        return count;
    }
}

/*
nums = [0, 1, 4, 4, 5, 7], lower = 3, upper = 6
// first find two numbers that add up to val withing the range of lower and up
 

 [1, 2, 5, 7, 9] 11
 [5,5,5,7,7]

*/

File: 2699-count-the-number-of-fair-pairs/count-the-number-of-fair-pairs.cpp
class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        long long count = 0;
        for (int i = 0; i < n; i++) {
            auto s = lower_bound(nums.begin() + i + 1, nums.end(), lower - nums[i]) - nums.begin();
            auto e = upper_bound(nums.begin() + i + 1, nums.end(), upper - nums[i]) - nums.begin() - 1;

            if (s <= e) {
                count += (e - s + 1);
            }
        }

        return count;
    }
};

/*
[0,1,4,4,5,7] 

[1,2,5,7,9]

*/